File: sdk/go/mohawksdk/logic.go
package mohawksdk

const round6Factor = 1_000_000

type FLState struct {
	Round   int
	Global  *float64
	Updates []float64
}

func round6(v float64) float64 {
	return float64(int64(v*round6Factor+0.5)) / round6Factor
}

func SwipScale(value float64) float64 {
	return round6(value * 1.5)
}

func FlApplyUpdate(state FLState, value float64) FLState {
	numUpdates := len(state.Updates) + 1

	if numUpdates >= 2 {
		sum := value
		for _, v := range state.Updates {
			sum += v
		}
		avg := round6(sum / float64(numUpdates))
		return FLState{Round: state.Round + 1, Global: &avg, Updates: nil}
	}

	return FLState{Round: state.Round, Updates: []float64{round6(value)}}
}

File: sdk/python/mohawk_sdk/logic.py
from typing import Dict, Any


def _round6(value: float) -> float:
    return round(float(value), 6)


def swip_scale(value: float) -> float:
    return _round6(float(value) * 1.5)


def fl_apply_update(state: Dict[str, Any], value: float) -> Dict[str, Any]:
    round_num = int(state.get("round", 0))
    updates = [float(v) for v in state.get("updates", [])]
    updates.append(float(value))

    if len(updates) >= 2:
        avg = sum(updates) / len(updates)
        return {
            "round": round_num + 1,
            "global": _round6(avg),
            "updates": [],
        }

    return {
        "round": round_num,
        "updates": [_round6(v) for v in updates],
    }
